accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,221 @@
+
+```cs
+// Example showing unshrouding a SensitiveData<char> into a temp char[].
+SensitiveData<char> sensitiveData = GetSensitiveStringData();
+char[] tempArray = GC.AllocateArray<char>(sensitiveData.Length, pinned: true);
+try
+{
+    sensitiveData.CopyTo(tempArray);
+    // use 'tempArray' here
+}
+finally
+{
+    Array.Clear(tempArray);
+}
+
+// Example showing unshrouding a SensitiveData<char> into a string.
+SensitiveData<char> sensitiveData = GetSensitiveStringData();
+string plaintextString = string.Create(
+    sensitiveData.Length, sensitiveData, (span, sensitiveData) =>
+    {
+        sensitiveData.CopyTo(span);
+    });
+```
+
+We would clearly document that this type is not intended to offer cryptographic protection of the buffer contents; it's simply intended to enforce more hygienic usage of data by callers.
+
+The type also intentionally does not give callers access to the raw backing span. This is because how the backing data is stored is strictly an implementation detail (it might not even be in-proc!), and exposing the span would leak that detail into the public API surface. The `CopyTo` method is the only way to get the raw contents. This also means that such buffers are immutable once created.
+
+__We should take care not to position this "sensitive buffer" type as a security feature for the purpose of achieving compliance.__ It's intended solely for code hygiene purposes and to avoid accidental disclosure. Scenarios like achieving PCI compliance require more fundamental runtime work, such as that under consideration at https://github.com/dotnet/runtime/issues/10480.
+
+This proposal intentionally excludes `ToString`, `ToArray`, or other methods which might unshroud the buffer. `ToString` is called frequently by library code as part of logging, and we always want unshrouding to be a _deliberate_ action. Additionally, we want the caller to maintain full control over the unshrouded buffer where the contents are written, which precludes "opinionated" APIs like `ToArray` which create such buffers themselves.
+
+### Best practices for implementation
+
+If the backing buffer for the sensitive data is stored in-proc, we should attempt to use a separate unmanaged heap for storing these values, and we should not hand out direct references to this heap. This helps mitigate the possibility of sensitive data being leaked due to corruption of the normal heap. This concept is [already in use by OpenSSL](https://www.openssl.org/docs/man1.1.0/man3/CRYPTO_secure_malloc_initialized.html) and a few high-sensitivity Windows services. If using a separate heap is not feasible, we should attempt to use .NET's pinned object heap.
```

Comment:
I don't think you've specified what Dispose() does, but I assume it zeros this fixed memory as well as deallocates it.

Reply:
> I don't think you've specified what Dispose() does

That's intentional. ;)

It releases the underlying resources. It may also take whatever steps are reasonable to prevent inadvertent disclosure of this data in the future, but those steps may depend on OS, runtime version, what the _T_ is (if we remove the _unmanaged_ constraint), etc.

Reply:
If you're going to note here that ideally it would use the POH, eg., it seems reasonable to note here that ideally it zeros the memory on the way out. But just a nit.